In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plot
dataset = pd.read_csv("heart.csv")
dataset

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
0     40   M           ATA        140          289          0     Normal   
1     49   F           NAP        160          180          0     Normal   
2     37   M           ATA        130          283          0         ST   
3     48   F           ASY        138          214          0     Normal   
4     54   M           NAP        150          195          0     Normal   
..   ...  ..           ...        ...          ...        ...        ...   
913   45   M            TA        110          264          0     Normal   
914   68   M           ASY        144          193          1     Normal   
915   57   M           ASY        130          131          0     Normal   
916   57   F           ATA        130          236          0        LVH   
917   38   M           NAP        138          175          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0      172              N      0.0       Up             0  
1      156              N      1.0     Flat             1  
2       98              N      0.0       Up             0  
3      108              Y      1.5     Flat             1  
4      122              N      0.0       Up             0  
..     ...            ...      ...      ...           ...  
913    132              N      1.2     Flat             1  
914    141              N      3.4     Flat             1  
915    115              Y      1.2     Flat             1  
916    174              N      0.0     Flat             1  
917    173              N      0.0       Up             0  

[918 rows x 12 columns]

In [4]:
#finding the quatitative columns for covariant calculation

quan=[]
qual=[]
for columnName in dataset.columns:
    if dataset[columnName].dtypes=='O':
        qual.append(columnName)
    else:
        quan.append(columnName)

In [5]:
quan

['Age',
 'RestingBP',
 'Cholesterol',
 'FastingBS',
 'MaxHR',
 'Oldpeak',
 'HeartDisease']

## Covariant on the numerical column
Covariance is a way to measure how two variables change together. If two variables have a positive covariance, they tend to increase or decrease together. A negative covariance means that when one variable increases, the other tends to decrease. 

In [6]:
dataset[quan].cov()

Age   RestingBP   Cholesterol  FastingBS       MaxHR  \
Age           88.974254   44.427519    -98.309701   0.790260  -91.750920   
RestingBP     44.427519  342.773903    204.323837   0.549778  -52.857808   
Cholesterol  -98.309701  204.323837  11964.891079 -12.076452  656.671673   
FastingBS      0.790260    0.549778    -12.076452   0.178968   -1.415709   
MaxHR        -91.750920  -52.857808    656.671673  -1.415709  648.228614   
Oldpeak        2.601774    3.254307      5.850572   0.023778   -4.363589   
HeartDisease   1.323300    0.990808    -12.663226   0.056246   -5.071057   

               Oldpeak  HeartDisease  
Age           2.601774      1.323300  
RestingBP     3.254307      0.990808  
Cholesterol   5.850572    -12.663226  
FastingBS     0.023778      0.056246  
MaxHR        -4.363589     -5.071057  
Oldpeak       1.137572      0.214307  
HeartDisease  0.214307      0.247420

## Observarion:
examples
1. Age and RestingBP (44.43): Positive, indicating that as age increases, resting blood pressure also tends to increase slightly.
2. Age and Cholesterol (-98.31): Negative, so as age increases, cholesterol tends to decrease, but the large negative value suggests this relationship may vary widely.
3. Age and FastingBS (0.79): Positive but close to zero, indicating almost no relationship between age and fasting blood sugar.
4. RestingBP and Cholesterol (204.32): Positive, suggesting that higher resting blood pressure is associated with higher cholesterol.
5. RestingBP and HeartDisease (0.99): Positive but low, showing that people with higher resting blood pressure might have a slightly higher chance of heart disease.



## Correlation on the numerical column
showing how strongly each pair of variables is related on a scale from -1 to 1. 
1:  Perfect positive correlation; when one variable increases, the other also increases proportionally.
-1: Perfect negative correlation; when one variable increases, the other decreases proportionally.
0: No correlation; the variables do not have any linear relationship.

In [7]:
dataset[quan].corr()

Age  RestingBP  Cholesterol  FastingBS     MaxHR   Oldpeak  \
Age           1.000000   0.254399    -0.095282   0.198039 -0.382045  0.258612   
RestingBP     0.254399   1.000000     0.100893   0.070193 -0.112135  0.164803   
Cholesterol  -0.095282   0.100893     1.000000  -0.260974  0.235792  0.050148   
FastingBS     0.198039   0.070193    -0.260974   1.000000 -0.131438  0.052698   
MaxHR        -0.382045  -0.112135     0.235792  -0.131438  1.000000 -0.160691   
Oldpeak       0.258612   0.164803     0.050148   0.052698 -0.160691  1.000000   
HeartDisease  0.282039   0.107589    -0.232741   0.267291 -0.400421  0.403951   

              HeartDisease  
Age               0.282039  
RestingBP         0.107589  
Cholesterol      -0.232741  
FastingBS         0.267291  
MaxHR            -0.400421  
Oldpeak           0.403951  
HeartDisease      1.000000

## Observarion:
example
1. Age and RestingBP (0.254): Small positive correlation; as age increases, resting blood pressure tends to increase slightly.
2. Age and MaxHR (-0.382): Moderate negative correlation; as age increases, maximum heart rate tends to decrease.
3. Cholesterol and MaxHR (0.236): Small positive correlation; higher cholesterol is weakly associated with higher max heart rate.
4. Cholesterol and HeartDisease (-0.233): Small negative correlation, meaning higher cholesterol is associated with slightly lower heart disease risk here, which might be counterintuitive and could depend on specific factors in this dataset.